### DISCUSSION TITLE
Crop production regression nodata bug?

### DISCUSSION THREAD

**Discussion by emlys** [natcap/invest#1302] [General]
Something seems off about the crop production model and interpretation of nodata. I'll try to explain with an example:

<img width="394" alt="image" src="https://user-images.githubusercontent.com/43770515/146622039-c9a05048-e5c0-45c0-ac15-266f8a68bd4a.png">

Considering the point X:
-> Barley may or may not be grown at point X (the climate bin map resolution is too low to be sure). 
-> Any barley grown at point X would most likely be in climate bin 70 (because all the nearest climate bin pixels are 70)
-> Any barley grown at point X should be modeled with the yield parameters for climate bin 70

-> It makes sense to estimate a lower yield for point X (accounting for the probability, scaling with distance from the known barley-growing pixels, that barley isn't grown there)
-> It doesn't make sense to lower all the yield parameters at point X. For example, the parameter `yield_ceiling` is the maximum barley yield, in tons/hectare, observed anywhere in climate bin 70. Why would the `yield_ceiling` at point X be lower than at a point farther to the right? That's like saying X is in a different climate bin altogether.

By interpolating all the yield parameters and treating nodata as 0 ([source](https://github.com/natcap/invest/blob/0604628a5b3e480ad0be55c78d5373adbbe1c217/src/natcap/invest/crop_production_regression.py#L391-L396)), we're effectively lowering all the yield parameters in that area.

The current model is effectively
`yield = (max_yield * p) * (1 - (b * p) * exp(-(c * p) * fertilization_rate))`
where `max_yield`, `b`, and `c` are all yield parameters tied to the climate bin. `p` is some value between 0 and 1 that represents parameter values being lowered when they're interpolated with a nodata area.

I suspect the correct model would be
`yield = p * max_yield * (1 - b_k2o * exp(-c_k20 * fertilization_rate))`
where `p` would be applied just once, at the end. Note that because of the exponent, these equations aren't equivalent.

Sorry if this makes no sense - I don't really understand the math/statistics involved, and I could be totally off.

**Comment by emlys** [comment]
Going to reach out to the model lead @jandrewjohnson about this. If we can find the paper published for this model, that might be enough to tell us if our math is right.

**Comment by jandrewjohnson** [comment]
This is an important implementation question that to my knowledge has not been addressed in the peer-reviewed sources. The regressions are defined to be valid for specific, discrete climate bins but there is no methodology discussed for what to do in locations where no climate bin is defined. I thought that, long ago, we had partially accounted for this by working with Jamie Gerber (from the original paper) to produce a map with no holes in the climate-bin map, but it seems this still has some areas where the NDV problem exists?

So, the solutions I can envision would be of 2 types:

1 - Recreate a climate bin map with no no-data spots.
2 - If that is not possible, your proposed equation change might make sense, but I'd want to know more about specifically where in the world this is happening. 

Can you provide more information on points 1-2?

**Comment by emlys** [comment]
Sorry for the delay @jandrewjohnson. I'm working on an example to illustrate this better for point 2.

As to point 1, there are large no-data areas and they look to be where the crop is not grown. For example, the climate bin map for banana has no-data in most of the northern hemisphere. The model converts no-data to 0, so the result would be the same as if the climate bin map no-data were replaced with 0.
